### PR TITLE
Limit the number of STHs produced to avoid client marking attacks.

### PR DESCRIPTION
**Pull request by benlaurie**
@robstradling

### CODE REVIEW

File: rfc6962-bis.xml [hunk 1]
```diff
@@ -1295,6 +1295,12 @@ but it is expected there will be a variety.
 	  <t hangText="Maximum Merge Delay">
 	    The MMD the log has committed to.
 	  </t>
+	  <t hangText="STH Frequency Period">
+	    The time, in seconds, over which <spanx style="verb">STH Frequency Count</spanx> is measured (see <xref target="STH"/>).
+	  </t>
+	  <t hangText="STH Frequency Count">
+	    The maximum number of STHs the log may produce in any <spanx style="verb">STH Frequency Period</spanx> (see <xref target="STH"/>).
+	  </t>
```
Comment (robstradling) [inline rfc6962-bis.xml]: Wouldn't it be simpler to define just one field - "Minimum Merge Delay" - instead of these two STH Frequency fields?

**Comment by mpalmer** [comment]
How will the contents of an STH be used to mark a client?  Isn't the only part of the STH that ever comes back to the server the number of entries?  If so, you can't get away with incrementing that without actually *adding* the entries, and a log which is adding entries quickly enough to be able to uniquely mark every client request is pretty quickly going to catch the attention of every monitor (if only because they'll have a tree size that is orders of magnitude larger than anyone else's).

File: rfc6962-bis.xml (hunk 1 above)
Comment (benlaurie) [inline rfc6962-bis.xml]: I wanted to allow some slop, rather than insist on a hard minimum.

**Comment by benlaurie** [comment]
@mpalmer This is true, but you still need to have some limit or you cannot objectively enforce a restriction.

File: rfc6962-bis.xml (hunk 1 above)
Comment (robstradling) [inline rfc6962-bis.xml]: OK.

Does the "Misbehaving Logs" section (currently 8.4) need to deal with the scenario of a log issuing more than "STH Frequency Count" STHs in any given "STH Frequency Period" ?

Comment (eranmes) [inline rfc6962-bis.xml]: I also support adding less fields - what do you think about omitting  the STH Frequency Period so the STH Frequency Count relates to the MMD? (i.e. how many STHs the log may produce in a given MMD-sized time frame).

**Comment by pphaneuf** [comment]
@mpalmer, the STH has a timestamp, a CT log could sign a new STH with the same tree size (and tree hash), but a different timestamp for every request, and remember who it handed each timestamp to. You'd then be able to who's sending STHs around...

**Comment by benlaurie** [comment]
@pphaneuf Good point - that could be revealed by, e.g., gossip.

File: rfc6962-bis.xml (hunk 1 above)
Comment (benlaurie) [inline rfc6962-bis.xml]: @eranmes I can go with that, or we can punt and just say "STH Frequency", since we don't actually define the metadata in detail.

Comment (benlaurie) [inline rfc6962-bis.xml]: @robstradling I don't think so - STH Frequency is a privacy, not security, consideration.

Comment (eranmes) [inline rfc6962-bis.xml]: @benlaurie I'm fine with either (I think that STH Count per MMD is more difficult to misinterpret, so easier to enforce when monitoring).

**Comment by benlaurie** [comment]
@eranmes PTAL

**Comment by ln5** [comment]
I think this is good.

Sadly enough, limiting the number of {tree head, tree size and timestamp} per time unit still leaves the attack described by @pphaneuf open because ECDSA signatures are typically non deterministic (i.e. changes for every signature mad over the same data). We'd need to mandate RFC6979 to stop that, see http://trac.tools.ietf.org/wg/trans/trac/ticket/83 .

I suppose that a mitigating action that could be taken by clients would be to "merge" STH:s with the same {head,size,timestamp} into one before sharing them through gossip.

**Comment by eranmes** [comment]
@ln5 Good point about signatures being non-deterministic. Would you like to bring it up on the list?

**Comment by benlaurie** [comment]
It is already in trac.

**Comment by benlaurie** [comment]
Merged at 239e6a04ba1cbd0d5a9681d905a8104c98db874d.

**Pull request closed without merging**